### PR TITLE
Add support for JSON and raw encodings

### PR DESCRIPTION
**Pull request by prashantv**
Refactor out encoding logic, and add support for JSON / Raw.

Here it is running against a simple YARPC server that's using JSON:
```
> ./yab -e json -p http://localhost:14376/moe hello
{
  "hostname": "tassadar.local",
  "language": "go",
  "service": "moe"
}
```

I'm planning to move the encoding-specific logic into a separate package in future.

@akshayjshah @abhinav

### CODE REVIEW

File: encoding.go [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+	errUnrecognizedEncoding = errors.New("unrecognized encoding, must be one of: json, thrift, raw")
+	errHealthThriftOnly     = errors.New("--health can only be used with Thrift")
+	errMissingMethodName    = errors.New("no method specified, specify --method method")
+)
+
+func (e Encoding) String() string {
+	return string(e)
+}
+
+// UnmarshalText imlements the encoding.TextUnmarshaler interface used by JSON, YAMl, etc.
```
Comment (abhinav) [inline encoding.go]: YAML

File: encoding.go [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+		e = Thrift
+	}
+
+	switch e {
+	case Thrift:
+		return newThriftEncoding(opts.ThriftFile, opts.MethodName)
+	case JSON:
+		return jsonEncoding{opts.MethodName}, nil
+	case Raw:
+		return rawEncoding{opts.MethodName}, nil
```
Comment (abhinav) [inline encoding.go]: These things are serializers, not encodings. Consider renaming the types.

File: encoding.go [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+		method, spec := getHealthSpec()
+		return thriftEncoding{method, spec}, nil
+	}
+
+	if opts.MethodName == "" {
+		return nil, errMissingMethodName
+	}
+
+	if e == UnspecifiedEncoding && strings.Contains(opts.MethodName, "::") {
+		e = Thrift
```
Comment (abhinav) [inline encoding.go]: nice

File: encoding.go [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+
+	// Request creates a transport.Request from the given []byte input.
+	Request(body []byte) (*transport.Request, error)
+
+	// Response converts a transport.Response into something that can be displayed to a user.
+	// For non-raw encodings, this is typically a map[string]interface{}.
+	Response(body *transport.Response) (interface{}, error)
+
+	// IsSuccess returns whether the given body is considered a successful response.
+	IsSuccess(body *transport.Response) error
```
Comment (abhinav) [inline encoding.go]: Name suggests it returns a boolean

File: encoding.go [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+	return JSON
+}
+
+func (e jsonEncoding) Request(input []byte) (*transport.Request, error) {
+	data, err := unmarshalJSONInput(input)
+	if err != nil {
+		return nil, err
+	}
+
+	bs, err := json.Marshal(data)
```
Comment (abhinav) [inline encoding.go]: I don't fully understand why you're decoding and re-encoding JSON? Is it just to check that it's valid JSON?

File: options.go [hunk 6]
```diff
@@ -31,11 +31,12 @@ type Options struct {
 
 // RequestOptions are request related options
 type RequestOptions struct {
-	ThriftFile  string `short:"t" long:"thrift" description:"Path of the .thrift file"`
-	MethodName  string `short:"m" long:"method" description:"The full Thrift method name (Svc::Method) to invoke"`
-	RequestJSON string `short:"r" long:"request" description:"The request body, in JSON format"`
-	RequestFile string `short:"f" long:"file" description:"Path of a file containing the request body in JSON"`
-	Health      bool   `long:"health" description:"Hit the health endpoint, Meta::health"`
+	Encoding    Encoding `short:"e" long:"encoding" description:"The encoding of the data. Defaults to Thrift if the method contains '::' or a thrift file is specified"`
```
Comment (abhinav) [inline options.go]: Nit: "Thrift file is specified"

Also, consider listing the options in the description.

File: thrift_request.go [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+package main
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/thriftrw/thriftrw-go/compile"
+	"github.com/yarpc/yab/sorted"
+	"github.com/yarpc/yab/thrift"
+	"github.com/yarpc/yab/transport"
```
Comment (abhinav) [inline thrift_request.go]: Import group ordering

**Comment by abhinav** [comment]
Minor nits. LGTM otherwise.

File: encoding.go (hunk 4 above)
Comment (prashantv) [inline encoding.go]: 👍 will rename to `CheckSuccess`

File: encoding.go (hunk 2 above)
Comment (prashantv) [inline encoding.go]: 👍 will rename.

File: encoding.go (hunk 5 above)
Comment (prashantv) [inline encoding.go]: Yeah just checking for valid JSON, and to send the same consistent JSON request (e.g. the ordering of fields etc is deterministic, whitespace is trimmed).

I'll add a comment.

File: thrift_request.go (hunk 7 above)
Comment (prashantv) [inline thrift_request.go]: fixed

File: encoding.go (hunk 1 above)
Comment (prashantv) [inline encoding.go]: fixed

File: options.go (hunk 6 above)
Comment (prashantv) [inline options.go]: Updated to include options + fix casing

File: encoding.go [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+type jsonSerializer struct {
+	methodName string
+}
+
+func (e jsonSerializer) Encoding() Encoding {
+	return JSON
+}
+
+// Request unmarshals the input to make sure it's valid JSON, and then
+// Marshals the map to produce consistent output with whiteespace removed
```
Comment (abhinav) [inline encoding.go]: whitespace

**Comment by breerly** [comment]
:balloon: :balloon: :balloon:

**Pull request merged**